Link object files that use `#[used]`

dianqk opened the pull request:
By directly linking the object files that use `#[used]`, we ensure the linker can see them.

This approach allows `#[used]` to avoid modifying symbol visibility, preserving local symbols. A similar example in C would be:

```c
// foo.c
__attribute__((constructor)) static void foo() {}
// main.c
void main(void) {}
```

If `foo.c` is placed in a static library, it will never be loaded unless the entire static library is fully loaded by `--whole-archive`.

This pull request removes some of the symbols in `symbols.o`. We can remove more symbols in a follow-up PR.

madsmtm committed:
Add a test case for `#[used]` with archive

dianqk committed:
Link object files that use `#[used]`

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13474318317/job/37651571016) [(plain)](https://github.com/rust-lang/rust/commit/b5d92a27ae9d2a25c3338b853a23876baa73a5e7/checks/37651571016/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04
ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.876 Building wheels for collected packages: reuse
#12 2.877   Building wheel for reuse (pyproject.toml): started
#12 3.088   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.089   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=0c2fd2aaf7b0bf8d6e131220aff14712a774c2ca462f3204d25460cbcf610b63
#12 3.090   Stored in directory: /tmp/pip-ephem-wheel-cache-vcokt54a/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.092 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.491 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.491 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.032 Collecting virtualenv
#12 4.032 Collecting virtualenv
#12 4.087   Downloading virtualenv-20.29.2-py3-none-any.whl (4.3 MB)
#12 4.241      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 28.2 MB/s eta 0:00:00
#12 4.300 Collecting platformdirs<5,>=3.9.1
#12 4.307   Downloading platformdirs-4.3.6-py3-none-any.whl (18 kB)
#12 4.347 Collecting filelock<4,>=3.12.2
#12 4.377 Collecting distlib<1,>=0.3.7
#12 4.385   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.396      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 49.4 MB/s eta 0:00:00
#12 4.396      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 49.4 MB/s eta 0:00:00
#12 4.475 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.656 Successfully installed distlib-0.3.9 filelock-3.17.0 platformdirs-4.3.6 virtualenv-20.29.2
#12 DONE 4.7s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      128960 kB
DirectMap2M:     5113856 kB
DirectMap1G:    13631488 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
downloading https://static.rust-lang.org/dist/2025-02-08/rustfmt-nightly-x86_64-unknown-linux-gnu.tar.xz
---
fmt check
fmt: checked 5853 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/compiler/rustc_codegen_ssa/src/back/linker.rs:1875: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
All checks passed!
checking python file formatting
28 files already formatted
checking C++ file formatting
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
  local time: Sat Feb 22 15:52:10 UTC 2025
  network time: Sat, 22 Feb 2025 15:52:10 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

DianQK commented:
Weird. The tests failed on macOS after I fixed various test cases. I'll look into it tomorrow.

bjorn3 commented:
I think we should link all rlibs with `--whole-archive`. That would also make some currently cgu partitioning dependent linker errors when an unused function references an undefined symbol consistently produce errors independent of which cgu it ends up in.

dianqk committed:
Ensuring proper codegen unit association for exported symbols.

dianqk committed:
No need to export C and compiler-builtins symbols for the executable

DianQK commented:
> Weird. The tests failed on macOS after I fixed various test cases. I'll look into it tomorrow.

I found that I switched from traversing the object file to using the symbol table. I'm unsure why rustc doesn't add the symbol table when building static libraries for macOS. I have implemented a more robust approach that doesn't rely on the static library.

BTW, I included an experimental commit that doesn't need to be part of the PR.

DianQK commented:
> I think we should link all rlibs with `--whole-archive`. That would also make some currently cgu partitioning dependent linker errors when an unused function references an undefined symbol consistently produce errors independent of which cgu it ends up in.

Beyond the overhead to compile time, this appears to subtly alter the semantics of rlibs during symbol resolution? In C, at least, we only typically load undefined symbols. I'm uncertain what impact this modification might have on real-world projects.

bjorn3 commented:
> Beyond the overhead to compile time, this appears to subtly alter the semantics of rlibs during symbol resolution? In C, at least, we only typically load undefined symbols. I'm uncertain what impact this modification might have on real-world projects.

You already can't depend on this behavior in Rust. Rustc changes the codegen unit partitioning between rustc versions, so if in one rustc version you wouldn't get a function pulled in, the next rustc version may cause it to be pulled in anyway if the function ends up co-located with another function that is used.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/13482765899/job/37670068308) [(plain)](https://github.com/rust-lang/rust/commit/bb7755467724233fd955bc431adb3df2123fc265/checks/37670068308/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
tests/pass/float_nan.rs ... ok
tests/pass/0weak_memory_consistency_sc.rs ... ok

FAILED TEST: tests/pass/alloc-access-tracking.rs
command: MIRI_ENV_VAR_TEST="0" MIRI_TEMP="/tmp/miri-uitest-2o6LBg" RUST_BACKTRACE="1" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/miri" "--error-format=json" "--sysroot=/checkout/obj/build/x86_64-unknown-linux-gnu/miri-sysroot" "-Dwarnings" "-Dunused" "-Ainternal_features" "-Zui-testing" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/miri_ui/0/tests/pass" "tests/pass/alloc-access-tracking.rs" "-Zmiri-track-alloc-id=21" "-Zmiri-track-alloc-accesses" "-Cpanic=abort" "--edition" "2021"
error: actual output differed from expected
Execute `./miri test --bless` to update `tests/pass/alloc-access-tracking.stderr` to the actual output
--- tests/pass/alloc-access-tracking.stderr
+++ <stderr output>
+++ <stderr output>
-note: tracking was triggered
-  --> tests/pass/alloc-access-tracking.rs:LL:CC
-   |
-LL |         let ptr = miri_alloc(123, 1);
-   |                   ^^^^^^^^^^^^^^^^^^ created Miri bare-metal heap allocation of 123 bytes (alignment ALIGN bytes) with id $ALLOC
-   = note: BACKTRACE:
-   = note: inside `miri_start` at tests/pass/alloc-access-tracking.rs:LL:CC
-
-note: tracking was triggered
-note: tracking was triggered
-  --> tests/pass/alloc-access-tracking.rs:LL:CC
-   |
-LL |         *ptr = 42; // Crucially, only a write is printed here, no read!
-   |         ^^^^^^^^^ write access to allocation with id $ALLOC
-   = note: BACKTRACE:
-   = note: inside `miri_start` at tests/pass/alloc-access-tracking.rs:LL:CC
-
-note: tracking was triggered
-note: tracking was triggered
-  --> tests/pass/alloc-access-tracking.rs:LL:CC
-   |
-LL |         assert_eq!(*ptr, 42);
-   |         ^^^^^^^^^^^^^^^^^^^^ read access to allocation with id $ALLOC
-   = note: BACKTRACE:
-   = note: inside `miri_start` at RUSTLIB/core/src/macros/mod.rs:LL:CC
-   = note: this note originates in the macro `assert_eq` (in Nightly builds, run with -Z macro-backtrace for more info)
-
-
-note: tracking was triggered
-  --> tests/pass/alloc-access-tracking.rs:LL:CC
-   |
-LL |         miri_dealloc(ptr, 123, 1);
-   |         ^^^^^^^^^^^^^^^^^^^^^^^^^ freed allocation with id $ALLOC
-   = note: BACKTRACE:
-   = note: inside `miri_start` at tests/pass/alloc-access-tracking.rs:LL:CC
-

---

Location:
   /cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ui_test-0.28.0/src/lib.rs:369

Backtrace omitted. Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.
error: test failed, to rerun pass `--test ui`
Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-145608eb94c4e144 --quiet` (exit status: 1)
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/x86_64-unknown-linux-gnu/release/deps/ui-145608eb94c4e144 --quiet` (exit status: 1)
Command has failed. Rerun with -v to see more details.
  local time: Sun Feb 23 12:30:02 UTC 2025
  network time: Sun, 23 Feb 2025 12:30:02 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

DianQK commented:
> I think we should link all rlibs with `--whole-archive`. That would also make some currently cgu partitioning dependent linker errors when an unused function references an undefined symbol consistently produce errors independent of which cgu it ends up in.

I'm testing this at #137481.

DianQK commented:
@bors try @rust-timer queue

bors commented:
:hourglass: Trying commit bb7755467724233fd955bc431adb3df2123fc265 with merge 46536f154414de0c4e28d0e9297a5ac10ef00174...
<!-- homu: {"type":"TryBuildStarted","head_sha":"bb7755467724233fd955bc431adb3df2123fc265","merge_sha":"46536f154414de0c4e28d0e9297a5ac10ef00174"} -->

rust-log-analyzer commented:
The job **`dist-x86_64-linux`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/13490818508/job/37688669808) [(plain)](https://github.com/rust-lang-ci/rust/commit/46536f154414de0c4e28d0e9297a5ac10ef00174/checks/37688669808/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
file:.git/config remote.origin.url=https://github.com/rust-lang-ci/rust
file:.git/config remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
file:.git/config gc.auto=0
file:.git/config http.https://github.com/.extraheader=AUTHORIZATION: basic ***
file:.git/config branch.try.remote=origin
file:.git/config branch.try.merge=refs/heads/try
file:.git/config remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
file:.git/config submodule.library/backtrace.active=true
file:.git/config submodule.library/backtrace.url=https://github.com/rust-lang/backtrace-rs.git
file:.git/config submodule.library/stdarch.active=true
---
[RUSTC-TIMING] rustc_borrowck test:false 114.850
   Compiling rustc_driver v0.0.0 (/checkout/compiler/rustc_driver)
error: linking with `clang` failed: exit status: 1
  |
  = note:  "clang" "-Wl,--version-script=/tmp/rustc17As2h/list" "-Wl,--no-undefined-version" "-m64" "/tmp/rustc17As2h/symbols.o" "<253 object files omitted>" "<sysroot>-rustc/x86_64-unknown-linux-gnu/release/deps/rustc_driver-774f694bdb7deb15.0c7313dw7euea77omqyi4yki5.rcgu.rmeta" "-Wl,--as-needed" "-Wl,-Bstatic" "/tmp/rustc17As2h/{librustc_codegen_llvm-6a81856ba97f4eba.rlib,librustc_llvm-8ed1c7633feca432.rlib,libblake3-c6c459bef0d0f74f.rlib,libpsm-e146f4bd642e37ac.rlib,libprofiler_builtins-c72ca7fe3dac98c5.rlib}.rlib" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib/{libcompiler_builtins-*}.rlib" "-Wl,-Bdynamic" "-lLLVM-20-rust-1.87.0-nightly" "-ldl" "-lgcc_s" "-lutil" "-lrt" "-lpthread" "-lm" "-ldl" "-lc" "-B<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/bin/gcc-ld" "-fuse-ld=lld" "-Wl,--eh-frame-hdr" "-Wl,-z,noexecstack" "-L" "<sysroot>-rustc/x86_64-unknown-linux-gnu/release/build/psm-bee8a26e2991f023/out" "-L" "<sysroot>-rustc/x86_64-unknown-linux-gnu/release/build/blake3-73e53664fd4755c1/out" "-L" "<sysroot>-rustc/x86_64-unknown-linux-gnu/release/build/blake3-73e53664fd4755c1/out" "-L" "<sysroot>-rustc/x86_64-unknown-linux-gnu/release/build/rustc_llvm-2f758f0fe20936e5/out" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/llvm/lib" "-L" "/rustroot/bin/../lib/gcc/x86_64-pc-linux-gnu/9.5.0/../../../../lib64" "-L" "<sysroot>/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-o" "<sysroot>-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_driver-774f694bdb7deb15.so" "-shared" "-Wl,-soname=librustc_driver-774f694bdb7deb15.so" "-Wl,-z,relro,-z,now" "-Wl,-O1" "-nodefaultlibs" "-u" "__llvm_profile_runtime" "-Wl,-z,origin" "-Wl,-rpath,$ORIGIN/../lib" "-Wl,--icf=all"
  = note: some arguments are omitted. use `--verbose` to show all linker arguments
  = note: rust-lld: error: version script assignment of 'global' to symbol '__addtf3' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__ashlti3' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__ashrti3' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__bswapdi2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__bswapsi2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__clzdi2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__clzti2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__ctzdi2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__ctzti2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__divmodti4' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__divtf3' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__eqtf2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__extenddftf2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__extendsftf2' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__fixsfti' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__fixtfdi' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__fixtfsi' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__fixtfti' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__fixunsdfdi' failed: symbol not defined
          rust-lld: error: version script assignment of 'global' to symbol '__fixunssfdi' failed: symbol not defined
          rust-lld: error: too many errors emitted, stopping now (use --error-limit=0 to see all errors)
          clang: error: linker command failed with exit code 1 (use -v to see invocation)

[RUSTC-TIMING] rustc_driver test:false 103.696
error: could not compile `rustc_driver` (lib) due to 1 previous error
Build completed unsuccessfully in 0:13:49
---
Caused by:
    Command RUST_BACKTRACE=full python3 /checkout/x.py build --target x86_64-unknown-linux-gnu --host x86_64-unknown-linux-gnu --stage 2 library/std --rust-profile-generate /tmp/tmp-multistage/opt-artifacts/rustc-pgo --set llvm.thin-lto=false --set llvm.link-shared=true [at /checkout/obj] has failed with exit code Some(1)

Stack backtrace:
   0: <anyhow::Error>::msg::<alloc::string::String>
             at /rust/deps/anyhow-1.0.95/src/backtrace.rs:27:14
   1: <opt_dist::exec::CmdBuilder>::run
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/exec.rs:80:17
   2: <opt_dist::exec::Bootstrap>::run
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/exec.rs:181:9
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/main.rs:222:13
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/main.rs:222:13
   4: <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}::{closure#0}, ()>
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/timer.rs:111:22
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/main.rs:211:9
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/main.rs:211:9
   6: <opt_dist::timer::TimerSection>::section::<opt_dist::execute_pipeline::{closure#1}, opt_dist::training::RustcPGOProfile>
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/timer.rs:111:22
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/main.rs:208:29
   8: opt_dist::main
             at /rustc/46536f154414de0c4e28d0e9297a5ac10ef00174/src/tools/opt-dist/src/main.rs:399:18
   9: <fn() -> core::result::Result<(), anyhow::Error> as core::ops::function::FnOnce<()>>::call_once
   9: <fn() -> core::result::Result<(), anyhow::Error> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/f0cb41030579cd1a6f72bd23f38e677052d5d485/library/core/src/ops/function.rs:250:5
  10: std::sys::backtrace::__rust_begin_short_backtrace::<fn() -> core::result::Result<(), anyhow::Error>, core::result::Result<(), anyhow::Error>>
             at /rustc/f0cb41030579cd1a6f72bd23f38e677052d5d485/library/std/src/sys/backtrace.rs:152:18
  11: std::rt::lang_start::<core::result::Result<(), anyhow::Error>>::{closure#0}
             at /rustc/f0cb41030579cd1a6f72bd23f38e677052d5d485/library/std/src/rt.rs:199:18
  12: core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once
  13: std::panicking::try::do_call
             at /rustc/f0cb41030579cd1a6f72bd23f38e677052d5d485/library/std/src/panicking.rs:587:40
  14: std::panicking::try
             at /rustc/f0cb41030579cd1a6f72bd23f38e677052d5d485/library/std/src/panicking.rs:550:19
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13490818508/job/37689233517)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/13490818508/job/37689233517","builder_name":"checks-actions"} -->

bjorn3 commented:
> > > Beyond the overhead to compile time, this appears to subtly alter the semantics of rlibs during symbol resolution? In C, at least, we only typically load undefined symbols. I'm uncertain what impact this modification might have on real-world projects.
> > 
> > 
> > You already can't depend on this behavior in Rust. Rustc changes the codegen unit partitioning between rustc versions, so if in one rustc version you wouldn't get a function pulled in, the next rustc version may cause it to be pulled in anyway if the function ends up co-located with another function that is used.
> 
> IIUC, if all libs are well-defined, no issue here? So this is a bad experience when encountering an undefined symbol error?

Correct. If all libs are well-defined, adding `--whole-archive` shouldn't cause any difference other than forcing `#[used]` statics to be kept even if they are in an object file that would otherwise not be pulled in. It won't grow the executable as we use `--gc-sections` anyway. When someone tries to use an undefined symbol this will however force the error to be reported consistently across rustc versions, making it less likely that someone opens a bug report with a linker "regression".

ChrisDenton commented:
If we do `--whole-archive` for all platforms then I think we should split Windows `raw-dylib` generated imports into a separate staticlib file. While it should now work with `whole-archive`, I still don't think it's a great idea.

bjorn3 commented:
> If we do --whole-archive for all platforms then I think we should split Windows raw-dylib generated imports into a separate staticlib file.

When creating a staticlib we need to keep it merged as for those there is no way to tell the build system to link a separate import library.

bors commented:
:umbrella: The latest upstream changes (presumably #135726) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).
